Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -243,6 +249,197 @@ extension Processor {
     _uncheckedForcedConsumeOne()
     return true
   }
+  
+  mutating func matchBuiltin(
+    _ cc: BuiltinCC,
+    _ isStrictAscii: Bool
+  ) -> Bool {
+    guard let c = load() else {
+      signalFailure()
+      return false
+    }
+
+    var matched: Bool
+    var next = input.index(after: currentPosition)
+    switch cc {
+      // lily note: when do these `any` cases appear? can they be compiled
+      // into consume instructions at compile time?
+    case .any, .anyGrapheme: matched = true
+    case .anyScalar:
+      matched = true
+      next = input.unicodeScalars.index(after: currentPosition)
+    case .digit:
+      matched = c.isNumber && (c.isASCII || !isStrictAscii)
+    case .hexDigit:
+      matched = c.isHexDigit && (c.isASCII || !isStrictAscii)
+    case .horizontalWhitespace:
+      matched = c.unicodeScalars.first?.isHorizontalWhitespace == true
+      && (c.isASCII || !isStrictAscii)
+    case .newlineSequence, .verticalWhitespace:
+      matched = c.unicodeScalars.first?.isNewline == true
+      && (c.isASCII || !isStrictAscii)
+    case .whitespace:
+      matched = c.isWhitespace && (c.isASCII || !isStrictAscii)
+    case .word:
+      matched = c.isWordCharacter && (c.isASCII || !isStrictAscii)
+    }
+    
+    if matched {
+      currentPosition = next
+      return true
+    } else {
+      signalFailure()
+      return false
+    }
+  }
+  
+  mutating func matchBuiltinScalar(
+    _ cc: BuiltinCC,
+    _ isStrictAscii: Bool
+  ) -> Bool {
+    guard let c = loadScalar() else {
+      signalFailure()
+      return false
+    }
+
+    var matched: Bool
+    var next = input.unicodeScalars.index(after: currentPosition)
+    switch cc {
+    case .any: matched = true
+    case .anyScalar: matched = true
+    case .anyGrapheme:
+      matched = true
+      next = input.index(after: currentPosition)
+    case .digit:
+      matched = c.properties.numericType != nil && (c.isASCII || !isStrictAscii)
+    case .hexDigit:
+      matched = Character(c).isHexDigit && (c.isASCII || !isStrictAscii)
+    case .horizontalWhitespace:
+      matched = c.isHorizontalWhitespace && (c.isASCII || !isStrictAscii)
+    case .verticalWhitespace:
+      matched = c.isNewline && (c.isASCII || !isStrictAscii)
+    case .newlineSequence:
+      matched = c.isNewline && (c.isASCII || !isStrictAscii)
```

Comment:
This was carried over from the old code, but why does this match a full `\r\n` sequence when in unicode scalars mode? Should we change that?

Reply:
It's what is specified in [the Unicode proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0363-unicode-for-string-processing.md):

> Additionally, `\R` and `CharacterClass.newline` provide a way to include the CR+LF pair, even when matching with Unicode scalar semantics.

I don't know the exact rationale behind it, but I would assume it's done in an effort to be compatible with other regex engines that match scalars by default but with `\R` still matching `\r\n`.